server/odc-service/src/main/java/com/oceanbase/odc/service/connection/logicaldatabase/core/executor/execution/model/ExecutionGroup.java

```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright (c) 2023 OceanBase.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.oceanbase.odc.service.connection.logicaldatabase.core.executor.execution.model;
+
+import java.util.List;
+import java.util.concurrent.ExecutorService;
+
+import lombok.Getter;
+import lombok.extern.slf4j.Slf4j;
+
+/**
+ * @Author: Lebie
+ * @Date: 2024/8/28 17:29
+ * @Description: []
+ */
+@Getter
+@Slf4j
+public abstract class ExecutionGroup<T, R> {
+    private final List<ExecutionUnit<T, R>> executionUnits;
+    private final List<ExecutionSubGroup<T, R>> subGroups;
+
+    public ExecutionGroup(List<ExecutionUnit<T, R>> executionUnits) {
+        this.executionUnits = executionUnits;
```

Comment:
full copy input of collections is a good hobbit, cause we'll not concern about the mutate of the input collection. In other words, it become thread local

Reply:
👍🏻 I modified it to `Collections.unmodifiableList(new ArrayList<>(executionUnits));`